finality-aleph/src/network/manager/compatibility.rs

```diff
@@ -0,0 +1,182 @@
+use std::{
+    fmt::{Display, Error as FmtError, Formatter},
+    mem::size_of,
+};
+
+use codec::{Decode, Encode, Error as CodecError, Input as CodecInput};
+
+use crate::network::{
+    manager::{DiscoveryMessage, NetworkData},
+    Data, Multiaddress,
+};
+
+type Version = u16;
+type ByteCount = u16;
+
+// We allow sending authentications of size up to 16KiB, that should be enough.
+const MAX_AUTHENTICATION_SIZE: u16 = 16 * 1024;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum VersionedAuthentication<M: Multiaddress> {
+    // Most likely from the future.
+    Other(Version, Vec<u8>),
+    V1(DiscoveryMessage<M>),
+}
+
+impl<D: Data, M: Multiaddress> TryInto<NetworkData<D, M>> for VersionedAuthentication<M> {
+    type Error = Error;
+
+    fn try_into(self) -> Result<NetworkData<D, M>, Self::Error> {
+        use VersionedAuthentication::*;
+        match self {
+            V1(message) => Ok(NetworkData::Meta(message)),
+            Other(v, _) => Err(Error::UnknownVersion(v)),
+        }
+    }
+}
+
+impl<M: Multiaddress> From<DiscoveryMessage<M>> for VersionedAuthentication<M> {
+    fn from(message: DiscoveryMessage<M>) -> VersionedAuthentication<M> {
+        VersionedAuthentication::V1(message)
+    }
+}
+
+fn encode_with_version(version: Version, mut payload: Vec<u8>) -> Vec<u8> {
+    let mut result = version.encode();
+    // This will produce rubbish if we ever try encodings that have more than u32::MAX bytes.
+    let num_bytes = payload.len() as ByteCount;
+    result.append(&mut num_bytes.encode());
+    result.append(&mut payload);
```

Comment:
You have 3 allocations here and multiple cloning of the underlying vectors:
* allocate vector of length `size_of(Version)` for `result` 
* allocation for `num_bytes.encode()` and then cloning it into `result` -- which requires extending the `result` vector
* cloning `payload` into `result` which, again, requires extendning the `result` vector.


Since this will be happening for every protocol message, it has a huge impact on the overall performance.

A better approach would be something serde-like where you pass in a buffer (`&mut Vec<u8>`) as an argument from the top-level call to `encode` and pre-allocate the vector with the expected length of the data.

Reply:
You are right, I will create an `encode_to` implementation in seperate PR. Also now I noticed we have `as u16` here, which is very stupid and will also be removed in this PR